packages/core/fs/src/MemoryFS.js

```diff
@@ -0,0 +1,493 @@
+// @flow
+import type {FileSystem, FileOptions} from './types';
+import type {FilePath} from '@parcel/types';
+import path from 'path';
+import {Readable, Writable} from 'stream';
+import {registerSerializableClass} from '@parcel/utils';
+import packageJSON from '../package.json';
+import WorkerFarm from '@parcel/workers';
+import nullthrows from 'nullthrows';
+
+const instances = new Map();
+let id = 0;
+
+type FSHandle = (fn: string, args: any[]) => any;
+type SerializedMemoryFS = {
+  id: number,
+  handle?: FSHandle
+};
+
+export class MemoryFS implements FileSystem {
+  dirs: Map<FilePath, Directory>;
+  files: Map<FilePath, File>;
+  id: number;
+  handle: any;
+  constructor() {
+    this.dirs = new Map([['/', new Directory()]]);
+    this.files = new Map();
+    this.id = id++;
+    instances.set(this.id, this);
+  }
+
+  static deserialize(opts: SerializedMemoryFS) {
+    return (
+      instances.get(opts.id) || new WorkerFS(opts.id, nullthrows(opts.handle))
+    );
+  }
+
+  serialize(): SerializedMemoryFS {
+    if (!this.handle) {
+      this.handle = WorkerFarm.createReverseHandle(
+        async (fn: string, args: any[]) => {
+          // $FlowFixMe
+          return this[fn](...args);
+        }
+      );
+    }
+
+    return {
+      id: this.id,
+      handle: this.handle
+    };
+  }
+
+  cwd() {
+    return '/';
+  }
+
+  _normalizePath(filePath: FilePath): FilePath {
+    return path.resolve(this.cwd(), filePath);
+  }
+
+  async writeFile(
+    filePath: FilePath,
+    contents: Buffer | string,
+    options: ?FileOptions
+  ) {
+    filePath = this._normalizePath(filePath);
+    if (this.dirs.has(filePath)) {
+      throw new Error(`EISDIR: ${filePath} is a directory`);
+    }
+
+    let dir = path.dirname(filePath);
+    if (!this.dirs.has(dir)) {
+      throw new Error(`ENOENT: ${dir} does not exist`);
```

Comment:
Node errors have a `code` prop (`ENOENT` in this case). From the `fs.stat` docs: "In case of an error, the `err.code` will be one of Common System Errors." Do we want to stay compatible with NodeFS in that regard?

Reply:
Updated